File: Task3_2_GradeCalculator/Calculator.cs
namespace GradeWeb;

public record Marks(decimal? Activities, decimal? Project, decimal? Examination);
public record GradeResult(decimal Total, string Grade);

public static class Calculator
{
    public static Dictionary<string, string[]> Validate(Marks marks)
    {
        var errors = new Dictionary<string, string[]>();
        foreach (var (name, value) in new[] { ("Activities", marks.Activities), ("Project", marks.Project), ("Examination", marks.Examination) })
            if (value is null or < 0 or > 100)
                errors[name] = ["Enter a numeric mark from 0 to 100."];
        return errors;
    }

    public static GradeResult Calculate(Marks marks)
    {
        if (Validate(marks).Count != 0) throw new ArgumentException("Marks must be between 0 and 100.");
        var total = decimal.Round(marks.Activities!.Value * .20m + marks.Project!.Value * .40m + marks.Examination!.Value * .40m, 2, MidpointRounding.AwayFromZero);
        var grade = total switch { >= 80 => "High Distinction", >= 70 => "Distinction", >= 60 => "Credit", >= 50 => "Pass", _ => "Fail" };
        return new(total, grade);
    }
}

File: Task3_3_PHPGradeCalculator/calculator.php
<?php
declare(strict_types=1);

/** Validate raw form values before any arithmetic. */
function validate_marks(array $input): array
{
    $errors = [];
    foreach (['activities', 'project', 'examination'] as $field) {
        $value = $input[$field] ?? null;
        if (!is_scalar($value) || is_bool($value) || !is_numeric($value)
            || !is_finite((float)$value) || (float)$value < 0 || (float)$value > 100) {
            $errors[$field] = ucfirst($field) . ': enter a numeric mark from 0 to 100.';
        }
    }
    return $errors;
}

function calculate_grade(array $input): array
{
    if (validate_marks($input)) throw new InvalidArgumentException('Invalid marks.');
    $total = round((float)$input['activities'] * 0.2 + (float)$input['project'] * 0.4 + (float)$input['examination'] * 0.4, 2, PHP_ROUND_HALF_UP);
    $grade = match (true) {
        $total >= 80 => 'High Distinction',
        $total >= 70 => 'Distinction',
        $total >= 60 => 'Credit',
        $total >= 50 => 'Pass',
        default => 'Fail'
    };
    return ['total' => $total, 'grade' => $grade];
}

function escape_html(mixed $value): string
{
    return htmlspecialchars(is_scalar($value) ? (string)$value : '', ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8');
}
